File: algorithm/check_bracket.py
# 오직 괄호의 짝이 맞는지만 검사하는 알고리즘
# 올바른 식인지 아닌지는 관심없음

# 아래 테스트 케이스 모두 통과
TEST_CASE = [
    ("5 + 2", True),
    ("(5 + 2)", True),
    ("(5 + 2", False),
    ("5 + 2)", False),
    ("(5)", True),
    ("5)", False),
    ("(1 + 2) * 3", True),
    ("1 + 2) * 3", False),
    ("(1 + 2 * 3", False),
    ("2 + 3 * (2 + 3) / 4 * (2 + 3)", True),
    ("2 + 3 * (2 + 3 / 4 * (2 + 3)", False),
    ("(2 + 3) * 2 + 3) / 4 * (2 + 3)", False),
    ("3 * (4 + ((1 + 2) * 3))", True),
    ("3 * 4 + ((1 + 2) * 3))", False),
    ("3 * 4 + (1 + 2 * 3))", False),
    ("5)+(3", False),
    ("))((", False)
]

# 오직 괄호의 짝이 맞는지만 검사하는 함수
def check_braket(strings):
    stack = []
    for char in strings:
        if char == "(":
            stack.append(char)
        elif char == ")":
            if not stack:
                return False
            if stack and stack[-1] == "(":
                stack.pop()
                continue
            stack.append(char)
    if stack:
        return False
    return True

# 테스트 진행, 모두 True가 나와야 올바른 알고리즘
for case, expected_result in TEST_CASE:
    result = check_braket(case)
    print(f"{expected_result == result} ##### {case}")

"""
결과

True ##### 5 + 2
True ##### (5 + 2)
True ##### (5 + 2
True ##### 5 + 2)
True ##### (5)
True ##### 5)
True ##### (1 + 2) * 3
True ##### 1 + 2) * 3
True ##### (1 + 2 * 3
True ##### 2 + 3 * (2 + 3) / 4 * (2 + 3)
True ##### 2 + 3 * (2 + 3 / 4 * (2 + 3)
True ##### (2 + 3) * 2 + 3) / 4 * (2 + 3)
True ##### 3 * (4 + ((1 + 2) * 3))
True ##### 3 * 4 + ((1 + 2) * 3))
True ##### 3 * 4 + (1 + 2 * 3))
True ##### 5)+(3
True ##### ))((

Process finished with exit code 0
"""

File: algorithm/checkBracket.js
// 오직 괄호의 짝이 맞는지만 검사하는 알고리즘
// 올바른 식인지 아닌지는 관심없음

// 아래 테스트 케이스 모두 통과
// TEST_CASE[](입력받은 수식, 기댓값)
const TEST_CASE = [
  ["5 + 2", true],
  ["(5 + 2)", true],
  ["(5 + 2", false],
  ["5 + 2)", false],
  ["(5)", true],
  ["5)", false],
  ["(1 + 2) * 3", true],
  ["1 + 2) * 3", false],
  ["(1 + 2 * 3", false],
  ["2 + 3 * (2 + 3) / 4 * (2 + 3)", true],
  ["2 + 3 * (2 + 3 / 4 * (2 + 3)", false],
  ["(2 + 3) * 2 + 3) / 4 * (2 + 3)", false],
  ["3 * (4 + ((1 + 2) * 3))", true],
  ["3 * 4 + ((1 + 2) * 3))", false],
  ["3 * 4 + (1 + 2 * 3))", false],
  ["5)+(3", false],
  ["))((", false],
];

export function checkBracket(strings) {
  if (strings.length === 0) {
    return true;
  }

  const stack = [];
  for (const char of strings) {
    if (char === "(") {
      stack.push(char);
    } else if (char === ")") {
      if (stack.length === 0) {
        return false;
      }
      if (stack.length > 0 && stack[stack.length - 1] == "(") {
        stack.pop();
        continue;
      }
      stack.push(char);
    }
  }
  return stack.length > 0 ? false : true;
}

function test() {
  for (const [strings, expected] of TEST_CASE) {
    const result = checkBracket(strings);
    console.log(`${expected === result} #### ${strings}`);
  }
}

/*

true #### 5 + 2
true #### (5 + 2)
true #### (5 + 2
true #### 5 + 2)
true #### (5)
true #### 5)
true #### (1 + 2) * 3
true #### 1 + 2) * 3
true #### (1 + 2 * 3
true #### 2 + 3 * (2 + 3) / 4 * (2 + 3)
true #### 2 + 3 * (2 + 3 / 4 * (2 + 3)
true #### (2 + 3) * 2 + 3) / 4 * (2 + 3)
true #### 3 * (4 + ((1 + 2) * 3))
true #### 3 * 4 + ((1 + 2) * 3))
true #### 3 * 4 + (1 + 2 * 3))
true #### 5)+(3
true #### ))((
*/
